Different output with --jobs=1 and --jobs=2

Originally reported by: **Robert Spier (BitBucket: [robert_spier](http://bitbucket.org/robert_spier))**

---

Parallelism (--jobs) changes the output of pylint.

It's not just the order of the tests, --jobs=2 outputs 18468 lines of output compared to only 21 for --jobs=1.  pylint 1.3.1 reports no lint errors.

$ venv/bin/pylint --jobs=2 --rcfile=$PWD/pylintrc  app/codein app/melange app/soc app/summerofcode app/settings.py app/urls.py app/main.py tests pavement.py setup.py 2>&1 | head
************\* Module codein.callback
W: 17, 0: import missing `from __future__ import absolute_import` (no-absolute-import)
W: 18, 0: import missing `from __future__ import absolute_import` (no-absolute-import)
W: 19, 0: import missing `from __future__ import absolute_import` (no-absolute-import)
W: 20, 0: import missing `from __future__ import absolute_import` (no-absolute-import)
************\* Module codein.types
W: 17, 0: import missing `from __future__ import absolute_import` (no-absolute-import)
W: 18, 0: import missing `from __future__ import absolute_import` (no-absolute-import)
W: 20, 0: import missing `from __future__ import absolute_import` (no-absolute-import)
W: 21, 0: import missing `from __future__ import absolute_import` (no-absolute-import)

$ venv/bin/pylint --jobs=1 --rcfile=$PWD/pylintrc  app/codein app/melange app/soc app/summerofcode app/settings.py app/urls.py app/main.py tests pavement.py setup.py 2>&1 | head
************\* Module main
E: 46, 2: print statement used (print-statement)
E: 47, 2: print statement used (print-statement)
E: 48, 2: print statement used (print-statement)
E: 49, 2: print statement used (print-statement)
E: 50, 2: print statement used (print-statement)
************\* Module tests.test_utils
E:658, 8: print statement used (print-statement)
E:662,10: print statement used (print-statement)
E:667, 8: print statement used (print-statement)

This is with the current head of the pylint repository. (1889:e404dd4d6e37 tip)

The source code being linted is https://code.google.com/p/soc/

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/374

I guess there is nothing special about such files -- any similar lines in 2 files really:
```
$ cat bin/1.py
1
2
3
4
5
6
```
same for bin/2.py.
```
$ pylint bin/1.py bin/2.py 
...
bin/2.py:1:0: R0801: Similar lines in 2 files
==1:[0:6]
==2:[0:6]
1
2
3
4
5
6 (duplicate-code)
==1:[0:6]
==2:[0:6]
1
2
3
4
5
6 (duplicate-code)
```

I experience the same behaviour (Pylint `2.15.0`, Python `3.8.13`). This behaviour seems to affect any check for which context is spread across files (cyclical imports, duplicate code).

At first I thought this was because input files were checked across different parallel workers, and therefore important context from a secondary file (e.g. duplicate code) was outside the context of the parallel worker. I don't think this is the case though; even when files are processed on the same worker (but jobs>1), the same behaviour is observed.

I think this has more to do with the different ways in which files are handled between [serial](https://github.com/PyCQA/pylint/blob/main/pylint/lint/pylinter.py#L676-L691) and [parallel](https://github.com/PyCQA/pylint/blob/main/pylint/lint/pylinter.py#L664-L673) execution (determined by the number of jobs). For example, for files `foo` and `bar`, I observed the similarity linter running twice for each file in parallel, but once for all files in serial. Of course if the similarity linter lints only `foo`, it will miss code duplicated in `bar`.

Another example not yet mentioned is `invalid-name`. `test1.py`:

```python
'''
Docstring
'''

class Test:
    '''
    Test Class
    '''

    CLASS_ATTR = 5

```

`test2.py`:

```python
'''
Docstring
'''
from test1 import Test

x = Test()
x.CLASS_ATTR = 3

```

```
~ % pylint --jobs=0 test1.py test2.py

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

```
~ % pylint --jobs=1 test1.py test2.py
************* Module test2
test2.py:7:0: C0103: Attribute name "CLASS_ATTR" doesn't conform to snake_case naming style (invalid-name)

-------------------------------------------------------------------
Your code has been rated at 8.00/10 (previous run: 10.00/10, -2.00)
```

--- a/doc/whatsnew/fragments/374.bugfix
+++ b/doc/whatsnew/fragments/374.bugfix
@@ -1,0 +1,3 @@
+Support ``duplicate-code`` message when parallelizing with ``--jobs``.
+
+Closes #374

--- a/pylint/checkers/similar.py
+++ b/pylint/checkers/similar.py
@@ -889,8 +889,11 @@ class SimilarChecker(BaseRawFileChecker, Similar):
         """Reduces and recombines data into a format that we can report on.
 
         The partner function of get_map_data()
+
+        Calls self.close() to actually calculate and report duplicate code.
         """
         Similar.combine_mapreduce_data(self, linesets_collection=data)
+        self.close()
 
 
 def register(linter: PyLinter) -> None:

--- a/pylint/lint/parallel.py
+++ b/pylint/lint/parallel.py
@@ -130,7 +130,7 @@ def check_parallel(
     """Use the given linter to lint the files with given amount of workers (jobs).
 
     This splits the work filestream-by-filestream. If you need to do work across
-    multiple files, as in the similarity-checker, then implement the map/reduce mixin functionality.
+    multiple files, as in the similarity-checker, then implement the map/reduce functionality.
     """
     # The linter is inherited by all the pool's workers, i.e. the linter
     # is identical to the linter object here. This is required so that

--- a/tests/test_self.py
+++ b/tests/test_self.py
@@ -233,7 +233,6 @@ class TestRunTC:
                 join(HERE, "functional", "a", "arguments.py"),
             ],
             out=out,
-            # We expect similarities to fail and an error
             code=MSG_TYPES_STATUS["E"],
         )
         assert (

--- a/tests/test_similar.py
+++ b/tests/test_similar.py
@@ -82,6 +82,23 @@ class TestSimilarCodeChecker:
             expected_output=expected_output,
         )
 
+    @pytest.mark.needs_two_cores
+    def test_duplicate_code_parallel(self) -> None:
+        path = join(DATA, "raw_strings_all")
+        expected_output = "Similar lines in 2 files"
+        self._test_output(
+            [
+                path,
+                "--disable=all",
+                "--enable=duplicate-code",
+                "--ignore-imports=no",
+                "--ignore-signatures=no",
+                "--min-similarity-lines=4",
+                "--jobs=2",
+            ],
+            expected_output=expected_output,
+        )
+
     def test_duplicate_code_raw_strings_disable_file(self) -> None:
         """Tests disabling duplicate-code at the file level in a single file."""
         path = join(DATA, "raw_strings_disable_file")

